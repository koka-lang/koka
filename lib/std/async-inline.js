/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.
 
  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

var effCancel = "std/async/cancelable";

var _cancel_handler = $std_core._makeHandler1( effCancel, 
    function( _loc, _x, _k ) { return _k(_x); },
    function(_l1, _l2, _cont, _op, _resume , _k ) { return _resume(_l1, _l2, _cont, $std_core._unit_, _k);  }
);


function _remove( arr, elem ) {
  var idx = arr.indexOf(elem);
  if (idx>=0) arr.splice(idx,1); // modify in-place
}

function _parent_cframe( hctx ) {
  if (hctx==null) hctx = $std_core.capture_handler_context();
  var top  = hctx.htop;
  while (top >= 0 && hctx.hstack[top].optag !== effCancel) { top--; }
  return (top >= 0 ? hctx.hstack[top].loc1 : null);
}

function _create_cancelable( caction ) {
  // identity operations
  var cframe = { canceled:false, contexts:[], nested:[], parent:_parent_cframe() };
  function cancel()   { return _cancel(cframe); }
  function action(_k) { return caction( cancel, _k); }  
  return $std_core._tuple2_( 
    (function(){ 
      if (cframe.parent) _remove(cframe.parent.nested,cframe); 
    }),
    (function(_k){ 
      if (cframe.parent) cframe.parent.nested.push(cframe);
      return _cancel_handler( cframe, action, _k ); 
    })
  );
}

function throw_canceled() {
  $std_core.throw_1("Canceled",Cancel);
}

function _cancel( cframe ) {
  if (cframe.canceled) return;
  cframe.canceled = true;
  var contexts = cframe.contexts; 
  var nested   = cframe.nested;
  cframe.contexts = [];
  cframe.nested   = [];
  contexts.forEach( function(hctx) { 
    if (hctx != null) {
      setTimeout( function() {
        $std_core.invoke_in_handler_context(hctx, function(_) {
          throw_canceled(); // throw a Cancel exception in that context
        });
      }, 0);
    }
  });
  nested.forEach( function(nestedFrame) {
    _cancel(nestedFrame);
  });
}

function _check_canceled( hctx ) {
  if (hctx==null) hctx = $std_core.capture_handler_context();
  var top = hctx.htop;
  while (top >= 0) {
    var h = hctx.hstack[top];
    if (h.optag===effCancel && h.loc1.canceled) {
      throw_canceled();
    }
    top--;
  }
}

function _register_cancel( cback ) {
  cback.cframe = _parent_cframe(cback.hctx);
  if (cback.cframe) {
    cback.cframe.contexts.push(cback.hctx);
  }
}

function _unregister_cancel( cback ) {
  if (cback.cframe) {
    _remove( cback.cframe.contexts, cback.hctx );
    cback.cframe = null;
  }
}



// --------------------------------------------------------
// Support for asynchronous actions with effect handlers
// --------------------------------------------------------

var _sync_nesting = 0;
var _sync_blocked = null;

function _async_start_synchronous() {
  _sync_nesting++;
  if (_sync_blocked===null) _sync_blocked = [];
}

function _async_end_synchronous() {
  if (_sync_nesting===0) return;
  _sync_nesting--; 
  if (_sync_nesting===0) {
    // now invoke all queued callbacks; we schedule all through a timeout
    _sync_blocked.forEach( function(entry) {
      var cb = function() { return entry.cback.invoke(entry.args); };
      setTimeout(cb,0);
    });
    _sync_blocked = null;
  }
}

function _Callback( cb, haserr, nodup ) {
  var self = this;
  self.cb     = cb;
  self.haserr = haserr;
  self.nodup  = nodup;
  self.cframe = null;
  // save the handler context
  self.hctx = $std_core.capture_handler_context();
  // remember if the callback itself was constructed inside a sync block
  self.in_synchronous = (_sync_nesting > 0); 
  // check for cancelation
  _check_canceled(self.hctx);
  _register_cancel(self);
}

_Callback.prototype.invoke = function(args) {
  var self = this;
  // unregister cancelation
  _unregister_cancel(self);
  // check if we return while a 'synchronous' block was started
  if (!self.in_synchronous && _sync_nesting > 0) { // if so, put ourselves on the queue
    if (self.nodup) {
      // update in-place possible previous occurrence of this event if 'nodup' was set (for timers or mouse events etc.)
      for (var i = 0; i < _sync_blocked.length; i++) {
        if (_sync_blocked[i].cback == self) {
          _sync_blocked[i].args = args;
          return; // updated in place, we are done.
        }
      }        
    }
    // and push ourselves on the queue to be called later
    _sync_blocked.push( { cback: self, args: args } );
    return;
  }
  else {
    // restore the handler context and invoke the callback
    $std_core.invoke_in_handler_context( self.hctx, function(_) {
      if (self.haserr) {
        if (args[0] != null) throw args[0];
        args.shift();         
      }
      return self.cb.apply(self.cb,args);
    }, $std_core._unit_ );
  }
}

// Create a callback wrapper around 'cb'. 
// If 'haserr' is true, the first parameter of the new call back is considered an error object
// and thrown automatically if not null. If it is null, it is not passed to the callback `cb`.
// If `nodup` is true, multiple call backs that are queued due to synchronized blocks are not duplicated
// and only the last call back is queued (for timers, mouse events etc).
function _async_callback( cb, haserr, nodup ) {
  var cback = new _Callback(cb,haserr,nodup);
  return (function(){ return cback.invoke(Array.prototype.slice.call(arguments)); });
}

function _async_fork( action ) {
  setTimeout( _async_callback( function() {
    action($std_core.async_yield);
  }), 0 );  
}

function _async_all( actions, k ) {
  var n       = actions.length;
  var results = new Array(n);
  var exn;
  if (n<=0) return k(results);
  
  // all concurrent actions end with this continuation
  // only when all are finished, we resume the passed in continuation
  function allk(i,x,f) {
    results[i] = x;
    n--;
    if (n <= 0) {
      // if an exception was set, raise it now (or keep raising on multiple resume!)
      if (exn!==undefined) throw exn;
      // in case a branch exits more than once (due to amb for example)
      // note: this should never happen as multiple resumes can only happen after
      // at least one resumption but that cannot be past an asyncAll..
      for(i = 0; i < results.length; i++) {
        if (results[i] === undefined) return f(x); 
      }
      return k(results);
    }
    else {
      return f(x);
    }
  }

  for( var i = 0; i < n; i++) {
    (function(j){ 
      var cb = _async_callback( function(_) {
        return $std_core.$catch( function(k1) {
            return actions[j](k1);
          },
          function(_exn,_k2) {
            if (exn===undefined) exn = _exn;
            return allk(j, undefined, $std_core.async_yield);
          },
          function(x) { 
            return allk(j,x,$std_core.Result); 
          });
        }); 
      // by using setTimeout we can schedule all continuations for execution
      setTimeout(cb,0); 
    })(i);  // capture i by value
  }
  return $std_core.async_yield();
}
