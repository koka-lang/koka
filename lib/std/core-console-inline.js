/*---------------------------------------------------------------------------
  Copyright 2012 Microsoft Corporation.

  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/
var _trace = function(s) { };
var _trace_any = function(s,x) { _trace("" + s + x); };
var _print = function(s) { _trace(s); };

if (typeof console !== undefined && typeof console.log === "function") {
  _trace = function(s) {
    console.log(s);
  };
}

if (typeof console !== undefined && typeof console.info === "function") {
  _trace_any = function(s,x) {
    console.info(s,x);
  };
}

function _println(msg) {
  _print(msg + "\n");
}


/*------------------------------------------------
  Console for Node
------------------------------------------------*/
if (_host === "node") {
  _print = function(s) {
    process.stdout.write(s);
  };
}

/*------------------------------------------------
  Console for Browser
------------------------------------------------*/
if (_host === "browser") {
  (function(){
    var escapes = {
        '&': '&amp;', // & first!
        '<': '&lt;',
        '>': '&gt;',
        '\'': '&apos;',
        '"': '&quot;',
        '\n': '<br>',
        '\r': '',
    };
    var escapes_regex = new RegExp("[" + Object.keys(escapes).join("") + "]", "g");

    function html_escape(txt) {
      return txt.replace(escapes_regex, function (s) {
        var r = escapes[s];
        return (r ? r : "");
      });
    }

    function get_console() {
      var cons = document.getElementById("koka-console");
      if (cons==null) {
        cons = document.createElement("div");
        cons.id = "koka-console";
        cons.style.fontFamily = "Consolas,Monaco,'Ubuntu Mono','Droid Sans Mono','Source Code Pro',monospace"
        cons.style.fontSize = "12pt";
        cons.style.width = "99%";
        document.body.appendChild(cons);
      }
      if (cons.display == "none") return null;
      return cons;
    }

    var output = null;
    function get_console_out()
    {
      if (output) return output;

      output = document.getElementById("koka-console-out");
      if (!output) {
        var cons = get_console();
        if (!cons) return null;

        output = document.createElement("div");
        output.id = "koka-console-out";
        output.style.fontFamily = "Consolas,Monaco,'Ubuntu Mono','Droid Sans Mono','Source Code Pro',monospace"
        output.style.fontSize = "12pt";
        output.style.width = "99%";
        output.style.height = "30ex";
        output.style.border = "gray solid 1px";
        output.wrap="off";
        output.style.overflow = "auto";
        output.style.whiteSpace = "pre";
        output.style.padding = "2px";
        output.style.margin = "2px";
        output.style.paddingBottom = "4px";
        output.readOnly = true;
        cons.appendChild(output);
      }

      if (!output.print) {
        output.print_html = function(s) {
          output.innerHTML = output.innerHTML + s;
          // try to scroll to the end
          if (output.createTextRange) {
            output.createTextRange().scrollIntoView(false);
          }
          else if (output.scrollTop !== undefined) {
            output.scrollTop = output.scrollHeight;
          }
        };
        output.print = function(s) {
          output.print_html(html_escape(s));
        };
      }

      return output;
    }

    _print = function(s) {
      var out = get_console_out();
      if (out && out.print) {
        out.print(s);
      }
    };

    _print.print_html = function(s) {
      var out = get_console_out();
      if (out && out.print_html) {
        out.print_html(s);
      }
    };
  })();
}
