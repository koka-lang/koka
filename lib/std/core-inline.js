/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.

  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/
var _host = "unknown"

if (typeof window !== 'undefined' && window.document) {
  _host = "browser";
}
else if (typeof importScripts !== 'undefined') {
  _host = "webworker"
}
else if (typeof process !== undefined) {
  _host = "node"
}


/*------------------------------------------------
  Number formatting
------------------------------------------------*/

function _string_repeat(s,n) {
  if (n<=0)  return "";
  if (n===1) return s;
  if (n===2) return s+s;
  var res = "";
  while(n > 0) {
    if (n & 1) res += s;
    n >>>= 1;
    s += s;
  }
  return res;
}

function _trimzeros(s) {
  return s.replace(/\.?0+$/,"");
}

function _gformat(x,format) {
  var hex = /^[xX]([0-9]*)/.exec(format)
  if (hex) {
    var w = parseInt(hex[1]);
    var s = x.toString(16)
    if (format[0] == 'X') s = s.toUpperCase();
    return (s.length<w ? _string_repeat("0",w - s.length) + s : s );
  }
  var exp = /^[eE]([0-9]*)/.exec(format)
  if (exp) {
    var w = parseInt(exp[1]);
    return (w>0 && w<=20 ? x.toExponential(w) : x.toExponential());
  }
  var fix = /^[fF]([0-9]*)/.exec(format)
  if (fix) {
    var w = parseInt(fix[1]);
    return _trimzeros((w > 0 && w <= 20) ? x.toFixed(w) : x.toFixed(20));
  }
  var expfix = /^[gG]([0-9]*)/.exec(format)
  if (expfix) {
    var w = parseInt(expfix[1]);
    return (w>0&&w<=20 ? x.toPrecision(w) : x.toPrecision());
  }
  /* default */
  return x.toString();
}


function _double_show_exp(d,fractionDigits) {
  var s;
  if (fractionDigits < 0) {
    // use at most |fractionDigits|
    s = d.toExponential();
  }
  else {
    // use exactly |fractionDigits|.
    if (fractionDigits > 20) fractionDigits = 20;
    s = d.toExponential(fractionDigits);
  }
  return s.replace(/[eE][\+\-]?0+$/,"");
}

function _double_show_fixed(d, fractionDigits) {
  var dabs = (d < 0.0 ? -d : d);
  if (dabs < 1.0e-15 || dabs > 1.0e+21) {
    return _double_show_exp(d,fractionDigits);
  }
  if (fractionDigits < 0) {
    // use at most |fractionDigits|
    var s = d.toFixed(-fractionDigits);              // show at full precision
    var cap = /^([\-\+]?\d+)(\.\d+)$/.exec(s);
    if (!cap) return s;
    var frac = cap[2].substr(0,1 - fractionDigits);  // then cut off
    return cap[1] + frac.replace(/(?:\.|([1-9]))0+$/,"$1"); // remove trailing zeros
  }
  else {
    // use exactly fractionDigits
    if (fractionDigits > 20) fractionDigits = 20;
    return d.toFixed(fractionDigits);
  }
}

/*------------------------------------------------
  Exceptions
------------------------------------------------*/
function _exn_capture_stack(exn) {
  if ("captureStackTrace" in Error) {
    Error.captureStackTrace(exn,_InfoException);  // best on Node.js
  }
  else {
    exn.stack = (new Error()).stack; // in browsers
  }
  if (exn.stack==null) exn.stack = "";
  // strip off leaf functions from the stack trace
  exn.stack = exn.stack.replace(/\n\s*at (exn_exception|exception|(Object\.)?throw_1|Object\.error|exn_error_pattern|Object\.error_pattern|exn_error_range|Object\._vector_at)\b.*/g,"");
}

var _InfoException = (function() {
  __extends(_InfoException,Error);
  function _InfoException(message,info) {
    this.name    = (info && info._tag ? info._tag : 'std/core/exception');
    this.message = message || "unknown error";
    this.info    = info || $Error;
    _exn_capture_stack(this);
  }
  return _InfoException;
})();

// System exceptions cannot be caught but do respect finalizers
var _InfoSystemException = (function() {
  __extends(_InfoSystemException,$std_core._SystemException);
  function _InfoSystemException(message,info) {
    $std_core._SystemException.call(this,message);
    this.name    = (info && info._tag ? info._tag : 'std/core/exception');
    this.info    = info || $Error;
    _exn_capture_stack(this);
  }
  return _InfoSystemException;
})();

function exn_exception(msg,info) {
  if (info===Cancel) return new _InfoSystemException(msg,info)
                else return new _InfoException(msg,info);
}

function exn_stacktrace( exn ) {
  if (exn instanceof Error && typeof exn.stack === "string") {
    return exn.stack;
  }
  else {
    return "";
  }
}

function exn_error_pattern(loc,def) {
  throw new _InfoException( loc + (def ? ": " + def : "") + ": pattern match failure", Pattern(loc,def));
}

function _unsupported_external(msg) {
  throw new _InfoException(msg, $Error);
}

function exn_error_range() {
  throw new _InfoException( "index out of bounds", Range );
}

function exn_throw( exn ) {
  throw exn;
}


function exn_info( exn ) {
  //console.log("exn_info: " + exn.stack);
  if (exn instanceof _InfoException && exn.info != null) {
    return exn.info;
  }
  else if (exn instanceof _InfoSystemException && exn.info != null) {
    return exn.info;
  }
  else if (exn instanceof $std_core._FinalizeException) {
    return Finalize;
  }
  else if (exn instanceof RangeError) {
    return Range;
  }
  else if (exn instanceof AssertionError) {
    return Assert;
  }
  else if (exn instanceof Error && typeof exn.code === "string" ) {
    return System(exn.code);
  }
  else {
    return $Error;
  }
}

function exn_message( exn ) {
  if (exn==null) {
    return "invalid error";
  }
  if (typeof exn.get_message === "function") { // for FinalizeException
    var msg = exn.get_message();
    if (typeof msg === "string") return msg;
  }
  if (typeof exn.message === "string") {
    return exn.message;
  }
  else if (typeof exn === "string") {
    return exn;
  }
  else if (typeof exn.toString === "function") {
    return exn.toString();
  }
  else {
    return "Unknown error";
  };
}


/*------------------------------------------------
  32-bit integer operations
--------------------------------------------------*/

function _int32_multiply(x,y) {
  var xhi = (x >> 16) & 0xFFFF;
  var xlo = x & 0xFFFF;
  var yhi = (y >> 16) & 0xFFFF;
  var ylo = y & 0xFFFF;
  var hi  = ((xhi * ylo) + (xlo * yhi));
  return (((hi << 16) + (xlo * ylo))|0)
}

function _int32_cmod(x,y) {
  if (y === 0) throw "modulus of zero";
  return ((x%y)|0);
}

function _int32_cdiv(x,y) {
  if (y === 0) throw "division by zero";
  return ((x/y)|0);
}

/*------------------------------------------------
  list helpers
------------------------------------------------*/

// Create a list with from a vector in constant stack space
function _vlist(elems,tail) {
  var xs = tail || Nil;
  if (elems!=null && elems.length>0) {
    for(var i = elems.length - 1; i >= 0; i--) {
      var elem = elems[i];
      if (elem !== undefined) xs = Cons(elem,xs);
    }
  }
  return xs;
}

// Create an array from a list with constant stack space
function _unvlist(list) {
  var elems = [];
  while(list) {
    elems.push(list.head);
    list = list.tail;
  }
  return elems;
}

// Create a vector with a function initializer
function _vector(n, f) {
  if (n<=0) return [];
  var a = new Array(n);
  for(var i = 0; i < n; i++) {
    a[i] = f(i);
  }
  return a;
}

// Index a vector
function _vector_at( v, i ) {
  var j = _int_to_int32(i);
  var x = v[j];
  if (x === undefined) { exn_error_range(); }
  return x;
}


/*------------------------------------------------
  General javascript helpers
------------------------------------------------*/
// make a shallow copy
function _copy(obj) {
  if (typeof obj !== 'object') return obj;
  var value = obj.valueOf();
  if (obj != value) return new obj.constructor(value);
  var newobj = {};
  for( var prop in obj) newobj[prop] = obj[prop];
  return newobj;
}

// get the fields of an object
function _fields(obj) {
  var props = [];
  for (var prop in obj) props.push(prop);
  return props;
}

// Export module `mod` extended with `exp`. Modifies `exp` in place and assigns to mod
function _export(mod,exp) {
  for(var prop in mod) {
    if (exp[prop] === undefined) {
      exp[prop] = mod[prop];
    }
  }
  return exp;
}

/* assign here so inlined primitives are available in system.core itself */
$std_core = _export($std_core, {
            "_export": _export
            // primitive operations emitted by the compiler
            , "_int32_multiply": _int32_multiply
            , "_int32_cmod": _int32_cmod
            , "_int32_cdiv": _int32_cdiv
            , "vlist": _vlist
            , "_vector_at": _vector_at
            , "_unsupported_external": _unsupported_external
            // integer operations that will be inlined
            , "_int_string": _int_string
            , "_int_double": _int_double
            , "_int_to_int32": _int_to_int32
            , "_double_to_int32": _double_to_int32
            ,"_double_round": _double_round
            , "_int_to_double": _int_to_double
            , "_int_iszero": _int_iszero
            , "_int_isodd": _int_isodd
            , "_int_negate": _int_negate
            , "_int_abs": _int_abs
            , "_int_sign": _int_sign
            , "_int_add": _int_add
            , "_int_sub": _int_sub
            , "_int_mul": _int_mul
            , "_int_div": _int_div
            , "_int_mod": _int_mod
            , "_int_divmod": _int_divmod
            , "_int_compare": _int_compare
            , "_int_eq": _int_eq
            , "_int_ne": _int_ne
            , "_int_gt": _int_gt
            , "_int_ge": _int_ge
            , "_int_lt": _int_lt
            , "_int_le": _int_le
            });


/*------------------------------------------------
  double arithmetic
------------------------------------------------*/


var _double_trunc = Math.trunc || function(x){ return x - x%1; };


/*------------------------------------------------
  integer arithmetic
------------------------------------------------*/


// We represent integers as a regular number as long as it is within _min_precise and _max_precise.
// Outside that we use bigInt objects.
const _max_precise = 9007199254740991; // 2^53 -1
const _min_precise = -_max_precise;

const _max_int32 =  0x7FFFFFFF;
const _min_int32 = -0x80000000;

// is a number small?
function _is_small(x) {
  return (x >= _min_precise && x <= _max_precise);
}

// If a big integer becomes small again, convert to a number
function _unbig(x) {
  if (x.isSmall) {
    if (_is_small(x.value)) return x.value;
  }
  else if (x.value.length===1) {
    var v = x.value[0];
    if (x.sign) v = -v;
    if (_is_small(v)) return v;
  }
  return x;
}

// Round a double with rounding to even on a tie.
function _double_round(d) {
    var n = Math.round(d); // rounds to +Infinity on a tie
    return (n - d == 0.5 && n % 2 != 0 ? n - 1 : n);  // if it was a tie, and n is odd, decrement by 1
}

// create an int from a double.
function _int_double(x) {
  if (_is_small(x)) return _double_round(x);
  if (isFinite(x)) return bigInt(x);
  if (x===Infinity) return _max_int32;
  if (x===-Infinity) return _min_int32;
  return 0;
}

// create an int from a string.
function _int_string(s) {
  return _unbig(bigInt(s));
}


// Clamp a big integer into a 32 bit integer range.
function _int_to_int32(x) {
  const v = (_is_small(x) ? x : x.toJSNumber());
  if (v > _max_int32) return _max_int32;
  if (v < _min_int32) return _min_int32;
  return (v|0);
}

// Clamp a double into a 32 bit integer range.
function _double_to_int32(x) {
  if (x > _max_int32) return _max_int32;
  if (x < _min_int32) return _min_int32;
  if (isNaN(x)) return 0;
  return (x|0);
}


function _int_to_double(x) {
  return (_is_small(x) ? x : x.toJSNumber());
}


// Wrappers for basic integer operations.
// Generally we first perform the operation and check afterwards for "overflow",
// and back down to bigInt operations only when needed.

function _big_add(x,y)    { return _unbig(bigInt(x).add(bigInt(y))); }
function _big_sub(x,y)    { return _unbig(bigInt(x).subtract(bigInt(y))); }
function _big_mul(x,y)    { return _unbig(bigInt(x).multiply(bigInt(y))); }
function _big_cdivmod(x,y){ return bigInt(x).divmod(bigInt(y)); }
function _big_cdiv(x,y)   { return _unbig(bigInt(x).divide(bigInt(y))); }
function _big_cmod(x,y)   { return _unbig(bigInt(x).mod(bigInt(y))); }
function _big_compare(x,y){ return bigInt(x).compare(bigInt(y)); }
function _big_negate(x)   { return bigInt(x).negate(); }
function _big_inc(x)      { return _unbig(bigInt(x).next()); }
function _big_dec(x)      { return _unbig(bigInt(x).prev()); }

function _big_count_pow10(x)  { return bigInt(x).count_pow10(); }
function _big_count_digits(x) { return bigInt(x).count_digits(); }
function _big_div_pow10(x,n)  { return _unbig(bigInt(x).div_pow10(n)); }
function _big_mul_pow10(x,n)  { return _unbig(bigInt(x).mul_pow10(n)); }

function _big_pow(x,n) { return _unbig(bigInt(x).pow(n)); }

function _big_divmod(x,y) {
  const d  = bigInt(y);
  const qr = _big_cdivmod(x,d);
  var q = qr.quotient;
  var r = qr.remainder;
  if (r.isNegative()) {
    if (d.isPositive()) { q = q.prev(); r = r.add(d) }
                   else { q = q.next(); r = r.subtract(d) }
  }
  return _tuple2_(_unbig(q),_unbig(r));
}

function _int_negate(x) {
  const z = -x;
  return (_is_small(z) ? z : _big_negate(x));
}

function _int_iszero(x) {
  return (!isNaN(x) ? x===0 : x.isZero());
}

function _int_isodd(x) {
  return (!isNaN(x) ? (x&1)===1 : x.isOdd());
}


function _int_abs(x) {
  return (!isNaN(x) ? Math.abs(x) : x.abs());
}

function _int_add(x,y) {
  const z = x + y;
  return (_is_small(z) ? z : _big_add(x,y));
}

function _int_sub(x,y) {
  const z = x - y;
  return (_is_small(z) ? z : _big_sub(x,y));
}

function _int_mul(x,y) {
  const z = x * y;
  return (_is_small(z) ? z : _big_mul(x,y));
}

function _int_cdivmod(x,y) {
  const q = _double_trunc(x / y);
  if (!isNaN(q)) {
    return _tuple2_(q,(x%y));
  }
  else {
    const qr = _big_cdivmod(x,y);
    return _tuple2_(_unbig(qr.quotient),_unbig(qr.remainder));
  }
}

function _int_cdiv(x,y) {
  const q = _double_trunc(x / y);
  return (!isNaN(q) ? q : _big_cdiv(x,y));
}

function _int_cmod(x,y) {
  const r = (x % y);
  return (!isNaN(r) ? (r) : _big_cmod(x,y));
}

function _int_divmod(x,y) {
  if (_int_iszero(y)) return 0;
  var q = _double_trunc(x / y);
  if (!isNaN(q)) {
    var r = x%y;
    if (r<0) {
      if (y>0) { q = q - 1; r = r + y; }
          else { q = q + 1; r = r - y; }
    }
    return _tuple2_(q,r);
  }
  else return _big_divmod(x,y)
}

function _int_div(x,y) {
  if (_int_iszero(y)) return 0;
  const q = _double_trunc(x/y);
  if (!isNaN(q)) {
    const r = (x%y);
    return (r<0 ? (y>0 ? q-1 : q+1) : q);
  }
  else return _big_divmod(x,y).fst;
}

function _int_mod(x,y) {
  if (_int_iszero(y)) return 0;
  const r = (x%y);
  if (!isNaN(r)) {
    return (r<0 ? (y>0 ? r+y : r-y) : r);
  }
  else return _big_divmod(x,y).snd;
}


function _int_compare(x,y) {
  const d = x - y;
  if (!isNaN(d)) {
    return (d>0 ? Gt : (d<0 ? Lt : Eq));
  }
  else {
    const c = _big_compare(x,y);
    return (c>0 ? Gt : (c<0 ? Lt : Eq));
  }
}

function _int_sign(x) {
  if (!isNaN(x)) {
    return (x>0 ? Gt : (x<0 ? Lt : Eq));
  }
  else {
    return (x.isZero() ? Eq : (x.isPositive() ? Gt : Lt));
  }
}

function _int_eq(x,y)   { return _int_compare(x,y)===Eq; }
function _int_ne(x,y)   { return _int_compare(x,y)!==Eq; }
function _int_lt(x,y)   { return _int_compare(x,y)===Lt; }
function _int_le(x,y)   { return _int_compare(x,y)!==Gt; }
function _int_gt(x,y)   { return _int_compare(x,y)===Gt; }
function _int_ge(x,y)   { return _int_compare(x,y)!==Lt; }


function _int_pow(i,exp) {
	if (_is_small(i)) {
		var j = Math.pow(i);
		if (_is_small(j)) return j;
	}
	return _big_pow(i,exp);
}

function _int_count_pow10(i) {
  return (_is_small(i) ? bigInt._count_pow10(i) : _big_count_pow10(i) );
}

function _int_count_digits(i) {
  return (_is_small(i) ? bigInt._count_digits(i) : _big_count_digits(i) );
}

function _int_mul_pow10(i,n) {
  return (_is_small(i) && n <= 14 ? _int_mul(i,Math.pow(10,n)) : _big_mul_pow10(i,n) );
}

function _int_cdiv_pow10(i,n) {
  return (_is_small(i) && n <= 14 ? _int_cdiv(i,Math.pow(10,n)) : _big_div_pow10(i,n) );
}

function _int_showhex(x,upper) {
  const s = x.toString(16);
  return (upper ? s.toUpperCase() : s);
}

function _int_parse(s,hex) {
  if (s==="") return Nothing;
  const cappre  = /^([\-\+])?(0[xX])?(.*)$/.exec(s);
  const sdigits = cappre[3].toLowerCase();
  const sign    = cappre[1] || "";
  if (cappre[2]) hex = true;
  const rx = (hex ? /^[0-9a-f]+$/ :
                     /^[0-9]+(?:e\+?[0-9]+)?$/);
  const cap = rx.exec(sdigits);
  if (!cap) return Nothing;
  else if (hex) return Just(_unbig(bigInt(sign + sdigits,16)));
  else return Just(bigInt(sign + sdigits));
}
