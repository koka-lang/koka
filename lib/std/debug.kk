/*---------------------------------------------------------------------------
  Copyright 2012-2016 Microsoft Corporation.

  This is free software; you can redistribute it and/or modify it under the
  terms of the Apache License, Version 2.0. A copy of the License can be
  found in the file "license.txt" at the root of this distribution.
---------------------------------------------------------------------------*/

/* Debugging support

*/
module std/debug



public extern breakpoint() : ndet () {
  cs "System.Diagnostics.Debugger.Break"
  js inline "(function(){ debugger; })()"
}

public extern launch() : ndet () {
  cs "System.Diagnostics.Debugger.Launch"
  js inline "(function(){ debugger; })()"
}