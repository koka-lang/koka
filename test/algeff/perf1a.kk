import std/time

fun fib(n) {
  if (n<=1) then 1 else fib(n - 1) + fib(n - 2);
}

fun comp() { 
  fib(4)
}

effect state<s> {
  fun get()      : s
  fun put(i : s) : ()
}
 
fun count() {
  comp()
  val i = get()
  if (i == 0) then i else {
    put(i - 1)
    count()
  }
}

fun testH(i0) {
  handle(count)(i=i0) {
    return x -> x
    get()    -> resume(i,i)
    put(j)   -> resume((),j)
  }
}

effect linear state1<s> {
  fun get1()    : s
  fun put1(i:s) : ()
}

fun count1() {
  comp();
  val i = get1()
  if (i == 0) then i else {
    put1(i - 1)
    count1()
  }
}

fun testH1(i0) {
  handle(count1)(i=i0) {
    return x -> x
    get1()   -> resume(i,i)
    put1(j)  -> resume((),j)
  }
}

fun testD(i) {
  comp();
  if (i==0) then i else testD(i - 1)
}

fun testR(n) {
  var i := n;
  fun countR() {
    comp()
    if (i == 0) then i else {
      i := i - 1
      countR()
    }
  }
  countR()
}


val n = 2000000
fun main() {
  print-elapsed({testD(n)}, "direct         ")
  print-elapsed({testR(n)}, "recursive      ")
  print-elapsed({testH1(n)},"linear handler ")
  print-elapsed({testH(n)}, "regular handler")
}