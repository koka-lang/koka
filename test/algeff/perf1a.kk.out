[01;30moperator branch (.handler) get: resume tail
[00;39;49m[00m[01;30moperator branch (.handler) put: resume tail
[00;39;49m[00m[01;30moperator branch (.handler) get1: resume tail
[00;39;49m[00m[01;30moperator branch (.handler) put1: resume tail
[00;39;49m[00m.Op-get : forall<a> .op-get<a>
.Op-get1 : forall<a> .op-get1<a>
.Op-put : forall<a> (i : a) -> .op-put<a>
.Op-put1 : forall<a> (i : a) -> .op-put1<a>
.Ops-get : forall<a,b> (get : .op-get<a>) -> .ops-state<a,a>
.Ops-get1 : forall<a,b> (get1 : .op-get1<a>) -> .ops-state1<a,a>
.Ops-put : forall<a,b> (put : .op-put<a>) -> .ops-state<a,()>
.Ops-put1 : forall<a,b> (put1 : .op-put1<a>) -> .ops-state1<a,()>
.tag-state : string
.tag-state1 : string
comp : () -> div int
count : () -> <state<int>,div> int
count1 : () -> <div,state1<int>> int
fib : (int) -> div int
get : forall<a> () -> (state<a>) a
get1 : forall<a> () -> (state1<a>) a
main : () -> <div,console,ndet> int
n : int
put : forall<a> (i : a) -> (state<a>) ()
put1 : forall<a> (i : a) -> (state1<a>) ()
testD : (int) -> div int
testH : (i0 : int) -> div int
testH1 : (i0 : int) -> div int
testR : (n : int) -> div int